### PR TITLE
Extend directive command to allow choosing interfaces

### PR DESCRIPTION
**Pull request by WoutervdBrink**
- [ ] Added or updated tests
- [x] Added Docs for all relevant versions
- [x] Updated CHANGELOG.md

Continuation of #1240; adds extensive options for implementing methods and interfaces belonging to directive types.

**Changes**

The `lighthouse:directive` command now takes the `--type`, `--field` and `--argument` options to implement the respective interfaces when creating a directive. The method stubs and imports are correctly created.

Writing tests for generators has proven to be quite difficult, as you can't really mock the Filesystem API used by the base `GeneratorCommand`, so you'd be spamming generated directives throughout your code. However, I have ran the command lots of times with varying options and so far it has not yet given invalid PHP code. For reference, an empty directive generated by the command looks like this:

```php
<?php

namespace App\GraphQL\Directives;

use Nuwave\Lighthouse\Schema\Directives\BaseDirective;

class DoesNothingDirective extends BaseDirective 
{
    // TODO implement the directive https://lighthouse-php.com/master/custom-directives/getting-started.html
}
```

And a directive with all options enabled results in this:

<details><summary>Click to expand</summary>

```php
<?php

namespace App\GraphQL\Directives;

use Closure;
use GraphQL\Language\AST\FieldDefinitionNode;
use GraphQL\Language\AST\InputValueDefinitionNode;
use GraphQL\Language\AST\ObjectTypeDefinitionNode;
use GraphQL\Language\AST\TypeDefinitionNode;
use GraphQL\Language\AST\TypeExtensionNode;
use Nuwave\Lighthouse\Execution\Arguments\ArgumentSet;
use Nuwave\Lighthouse\Schema\AST\DocumentAST;
use Nuwave\Lighthouse\Schema\Directives\BaseDirective;
use Nuwave\Lighthouse\Schema\Values\FieldValue;
use Nuwave\Lighthouse\Schema\Values\TypeValue;
use Nuwave\Lighthouse\Support\Contracts\ArgBuilderDirective;
use Nuwave\Lighthouse\Support\Contracts\ArgDirectiveForArray;
use Nuwave\Lighthouse\Support\Contracts\ArgManipulator;
use Nuwave\Lighthouse\Support\Contracts\ArgResolver;
use Nuwave\Lighthouse\Support\Contracts\ArgTransformerDirective;
use Nuwave\Lighthouse\Support\Contracts\FieldManipulator;
use Nuwave\Lighthouse\Support\Contracts\FieldMiddleware;
use Nuwave\Lighthouse\Support\Contracts\FieldResolver;
use Nuwave\Lighthouse\Support\Contracts\TypeExtensionManipulator;
use Nuwave\Lighthouse\Support\Contracts\TypeManipulator;
use Nuwave\Lighthouse\Support\Contracts\TypeMiddleware;
use Nuwave\Lighthouse\Support\Contracts\TypeResolver;

class DoesManyThingsDirective extends BaseDirective implements TypeManipulator, TypeMiddleware, TypeResolver, TypeExtensionManipulator, FieldResolver, FieldMiddleware, FieldManipulator, ArgDirectiveForArray, ArgTransformerDirective, ArgBuilderDirective, ArgResolver, ArgManipulator
{
    // TODO implement the directive https://lighthouse-php.com/master/custom-directives/getting-started.html

    /**
     * Apply manipulations from a type definition node.
     *
     * @param  \Nuwave\Lighthouse\Schema\AST\DocumentAST  $documentAST
     * @param  \GraphQL\Language\AST\TypeDefinitionNode  $typeDefinition
     * @return void
     */
    public function manipulateTypeDefinition(DocumentAST &$documentAST, TypeDefinitionNode &$typeDefinition)
    {
        // TODO implement the type manipulator
    }

    /**
     * Handle a type AST as it is converted to an executable type.
     *
     * @param  \Nuwave\Lighthouse\Schema\Values\TypeValue  $value
     * @param  \Closure  $next
     * @return \GraphQL\Type\Definition\Type
     */
    public function handleNode(TypeValue $value, Closure $next)
    {
        // TODO implement the type middleware
    }

    /**
     * Resolve a type AST to a GraphQL Type.
     *
     * @param  \Nuwave\Lighthouse\Schema\Values\TypeValue  $value
     * @return \GraphQL\Type\Definition\Type
     */
    public function resolveNode(TypeValue $value)
    {
        // TODO implement the type resolver
    }

    /**
     * Apply manipulations from a type extension node.
     *
     * @param  \Nuwave\Lighthouse\Schema\AST\DocumentAST  $documentAST
     * @param  \GraphQL\Language\AST\TypeExtensionNode  $typeExtension
     * @return void
     */
    public function manipulateTypeExtension(DocumentAST &$documentAST, TypeExtensionNode &$typeExtension)
    {
        // TODO implement the type extension manipulator
    }

    /**
     * Set a field resolver on the FieldValue.
     *
     * This must call $fieldValue->setResolver() before returning
     * the FieldValue.
     *
     * @param  \Nuwave\Lighthouse\Schema\Values\FieldValue  $fieldValue
     * @return \Nuwave\Lighthouse\Schema\Values\FieldValue
     */
    public function resolveField(FieldValue $fieldValue)
    {
        // TODO implement the field resolver
    }

    /**
     * Wrap around the final field resolver.
     *
     * @param  \Nuwave\Lighthouse\Schema\Values\FieldValue  $fieldValue
     * @param  \Closure  $next
     * @return \Nuwave\Lighthouse\Schema\Values\FieldValue
     */
    public function handleField(FieldValue $fieldValue, Closure $next)
    {
        // TODO implement the field middleware
    }

    /**
     * Manipulate the AST based on a field definition.
     *
     * @param  \Nuwave\Lighthouse\Schema\AST\DocumentAST  $documentAST
     * @param  \GraphQL\Language\AST\FieldDefinitionNode  $fieldDefinition
     * @param  \GraphQL\Language\AST\ObjectTypeDefinitionNode  $parentType
     * @return void
     */
    public function manipulateFieldDefinition(
        DocumentAST &$documentAST,
        FieldDefinitionNode &$fieldDefinition,
        ObjectTypeDefinitionNode &$parentType
    )
    {
        // TODO implement the field manipulator
    }

    /**
     * Apply transformations on the value of an argument given to a field.
     *
     * @param  mixed  $argumentValue
     * @return mixed
     */
    public function transform($argumentValue)
    {
        // TODO implement the arg transformer
    }

    /**
     * Add additional constraints to the builder based on the given argument value.
     *
     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder  $builder
     * @param  mixed  $value
     * @return \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder
     */
    public function handleBuilder($builder, $value)
    {
        // TODO implement the arg builder
    }

    /**
     * @param  mixed  $root  The result of the parent resolver.
     * @param  mixed|ArgumentSet|ArgumentSet[]  $value  The slice of arguments that belongs to this nested resolver.
     * @return mixed
     */
    public function __invoke($root, $value)
    {
        // TODO implement the arg resolver
    }

    /**
     * Manipulate the AST.
     *
     * @param  \Nuwave\Lighthouse\Schema\AST\DocumentAST  $documentAST
     * @param  \GraphQL\Language\AST\InputValueDefinitionNode  $argDefinition
     * @param  \GraphQL\Language\AST\FieldDefinitionNode  $parentField
     * @param  \GraphQL\Language\AST\ObjectTypeDefinitionNode  $parentType
     * @return void
     */
    public function manipulateArgDefinition(
        DocumentAST &$documentAST,
        InputValueDefinitionNode &$argDefinition,
        FieldDefinitionNode &$parentField,
        ObjectTypeDefinitionNode &$parentType
    )
    {
        // TODO implement the arg manipulator
    }
}
```

</details>

**Breaking changes**

None :smile:

### CODE REVIEW

File: src/Console/DirectiveCommand.php [hunk 1]
```diff
@@ -25,16 +29,151 @@ class DirectiveCommand extends LighthouseGeneratorCommand
...
+     * The imports required by the various interfaces, if any.
+     *
+     * @var Collection
+     */
+    protected $imports;
+
     protected function getNameInput(): string
     {
-        return ucfirst($this->argument('name')).'Directive';
+        return ucfirst(trim($this->argument('name'))).'Directive';
```
Comment (spawnia) [inline src/Console/DirectiveCommand.php]: Did you find that calling `trim()` is really necessary? I think that the underlying command class already takes care of that.

File: src/Console/DirectiveCommand.php [hunk 2]
```diff
@@ -25,16 +29,151 @@ class DirectiveCommand extends LighthouseGeneratorCommand
...
 
+    /**
+     * Build the class with the given name.
+     *
+     * @param  string  $name
+     * @return string
+     *
+     * @throws \Illuminate\Contracts\Filesystem\FileNotFoundException
+     */
+    protected function buildClass($name)
```
Comment (spawnia) [inline src/Console/DirectiveCommand.php]:
```suggestion
    protected function buildClass(string $name): string
```

Always use the strictest possible type hints.

File: src/Console/DirectiveCommand.php [hunk 3]
```diff
@@ -25,16 +29,151 @@ class DirectiveCommand extends LighthouseGeneratorCommand
...
+        if ($this->option('type')) {
+            $this->askForInterfaces($stub, ['TypeManipulator', 'TypeMiddleware', 'TypeResolver', 'TypeExtensionManipulator']);
+        }
+
+        if ($this->option('field')) {
+            $this->askForInterfaces($stub, ['FieldResolver', 'FieldMiddleware', 'FieldManipulator']);
+        }
+
+        if ($this->option('argument')) {
+            // https://lighthouse-php.com/4.10/custom-directives/argument-directives.html
```
Comment (spawnia) [inline src/Console/DirectiveCommand.php]:
```suggestion
```

File: src/Console/DirectiveCommand.php [hunk 4]
```diff
@@ -25,16 +29,151 @@ class DirectiveCommand extends LighthouseGeneratorCommand
...
+            // Arg directives always either implement ArgDirective or ArgDirectiveForArray.
+            if ($this->confirm('Will your argument directive apply to a list of items?')) {
```
Comment (spawnia) [inline src/Console/DirectiveCommand.php]: I like how that part is presented to the user, nicely done.

File: src/Console/DirectiveCommand.php [hunk 5]
```diff
@@ -25,16 +29,151 @@ class DirectiveCommand extends LighthouseGeneratorCommand
...
+        $this->cleanup($stub);
+
+        return $stub;
+    }
+
+    /**
+     * Ask the user if the directive should implement any of the given
+     * interfaces.
+     *
+     * @param string $stub
```
Comment (spawnia) [inline src/Console/DirectiveCommand.php]:
```suggestion
     * @param  string  $stub
```

File: src/Console/DirectiveCommand.php [hunk 6]
```diff
@@ -25,16 +29,151 @@ class DirectiveCommand extends LighthouseGeneratorCommand
...
+     * @param array $interfaces
+     * @return void
+     */
+    protected function askForInterfaces(string &$stub, array $interfaces)
```
Comment (spawnia) [inline src/Console/DirectiveCommand.php]:
```suggestion
    protected function askForInterfaces(string &$stub, array $interfaces): void
```

File: src/Console/stubs/directives/arg_builder_directive.stub [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+     * Add additional constraints to the builder based on the given argument value.
+     *
+     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder  $builder
+     * @param  mixed  $value
+     * @return \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder
+     */
+    public function handleBuilder($builder, $value)
+    {
+        // TODO implement the arg builder
+    }
```
Comment (spawnia) [inline src/Console/stubs/directives/arg_builder_directive.stub]: Nit: having a newline here would be nice, it plays better with git and other tools.

File: src/Console/stubs/directives/arg_resolver.stub [hunk 8]
```diff
@@ -0,0 +1,9 @@
+    /**
+     * @param  mixed  $root  The result of the parent resolver.
+     * @param  mixed|ArgumentSet|ArgumentSet[]  $value  The slice of arguments that belongs to this nested resolver.
```
Comment (spawnia) [inline src/Console/stubs/directives/arg_resolver.stub]: Those should be FQCN according to the Laravel style guide.

**Comment by spawnia** [comment]
In terms of testing, it might be a good idea to isolate a pure functional core of this command that does not write to the filesystem.

File: src/Console/DirectiveCommand.php (hunk 1 above)
Comment (WoutervdBrink) [reply src/Console/DirectiveCommand.php]: You know what, I have no idea how that actually got there :thinking:... No, the trim is not necessary.

**Pull request merged**